parsers/ES.py

```diff
@@ -113,17 +159,61 @@ def check_valid_parameters(
 
 
 def fetch_island_data(
-    zone_key: ZoneKey, session: Session, target_datetime: Optional[datetime]
-) -> List[Response]:
+    zone_key: ZoneKey,
+    session: Session,
+    target_datetime: Optional[datetime],
+    data_mapping: Dict = DATA_MAPPING,
+):
     """Fetch data for the given zone key."""
+    tz = TIMEZONES_MAPPING[zone_key]
     if target_datetime is None:
-        date = target_datetime
+        date = utcnow().to(tz).format("YYYY-MM-DD")
     else:
         date = target_datetime.strftime("%Y-%m-%d")
-
-    data: List[Response] = ZONE_FUNCTION_MAP[zone_key](session).get_all(date)
-    if data:
-        return data
+    system = ZONE_MAPPING[zone_key]
+    zone = API_CODE_MAPPING[zone_key]
+    res = session.get(
+        f"https://demanda.ree.es/WSvisionaMoviles{system}Rest/resources/demandaGeneracion{system}?curva={zone}&fecha={date}"
+    )
+    if not res.ok:
+        raise ParserException(
+            "ES.py",
+            f"Failed fetching data for {zone_key}",
+            zone_key,
+        )
+    json = loads(res.text.replace("null(", "", 1).replace(r");", "", 1))
+    data = json["valoresHorariosGeneracion"]
+    responses = []
+    for value in data:
+        ts = value.pop("ts")
+        arrow = get(f"{ts} {tz}", "YYYY-MM-DD HH:mm ZZZ")
+        response = {}
+        response["timestamp"] = arrow
+        production = ProductionMix()
+        storage = StorageMix()
+        exchange = {}
+        for mode in value:
+            try:
+                if "production." in data_mapping[mode]:
+                    production.add_value(
+                        data_mapping[mode].replace("production.", ""), value[mode]
+                    )
+                elif "storage." in data_mapping[mode]:
+                    storage.add_value(
+                        data_mapping[mode].replace("storage.", ""), value[mode] * -1
+                    )
+                elif "exchange." in data_mapping[mode]:
+                    exchange[data_mapping[mode].replace("exchange.", "")] = value[mode]
+                elif data_mapping[mode] == "demand":
+                    response["demand"] = value[mode]
+            except KeyError:
+                raise ParserException("ES.py", f"Found unknown key:{mode}", mode)
+        response["production"] = production
+        response["storage"] = storage
+        response["exchange"] = exchange
+        responses.append(response)
+    if responses:
+        return responses
```

Comment:
This is very complicated comnpared to the old version. Espeically with all the string matching and replacing going on.